Race condition in DockerContainerTestCase

… causes the transit failure of various unit tests in the unit test suit due to a required docker container not publishing its ports promptly after creation.

The failure may present itself as …
```
======================================================================
ERROR: setUpClass (indexer.test_tdr.TestTDRHCAPlugin)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/achave11/Pycharm/Azul/azul.sc/.venv/lib/python3.11/site-packages/more_itertools/more.py", line 539, in one
    first_value = next(it)
                  ^^^^^^^^
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/achave11/Pycharm/Azul/azul.sc/test/indexer/test_tdr.py", line 190, in setUpClass
    cls.netloc = cls._create_container(image=image,
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/achave11/Pycharm/Azul/azul.sc/test/docker_container_test_case.py", line 94, in _create_container
    port = one(ports[f'{container_port}/tcp'])
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/achave11/Pycharm/Azul/azul.sc/.venv/lib/python3.11/site-packages/more_itertools/more.py", line 541, in one
    raise (
ValueError: too few items in iterable (expected 1)

----------------------------------------------------------------------
Ran 432 tests in 760.136s

FAILED (errors=1, skipped=3)
make: *** [test] Error 1
```

… a ValueError failure in any test class that inherits from DockerContainerTestCase. Alternatively, the ValueError is raised with the last call in the tree like `super().setUpClass()` followed by the ValueError, and completely omitting the `one(ports[f'{container_port}/tcp'])` part.

--- a/test/docker_container_test_case.py
+++ b/test/docker_container_test_case.py
@@ -1,4 +1,8 @@
+from datetime import (
+    datetime,
+)
 import os
+import time
 from typing import (
     ClassVar,
     Optional,
@@ -75,6 +79,7 @@ class DockerContainerTestCase(AzulUnitTestCase):
         log.info('Launching %scontainer from image %s',
                  'sibling ' if is_sibling else '', image)
         ports = None if is_sibling else {container_port: ('127.0.0.1', None)}
+        start = datetime.now()
         container = cls._docker.containers.run(image,
                                                detach=True,
                                                auto_remove=True,
@@ -82,21 +87,29 @@ class DockerContainerTestCase(AzulUnitTestCase):
                                                **kwargs)
         try:
             container_info = cls._docker.api.inspect_container(container.id)
-            network_settings = container_info['NetworkSettings']
             if is_sibling:  # no coverage
-                container_ip = network_settings['IPAddress']
+                container_ip = container_info['NetworkSettings']['IPAddress']
                 assert isinstance(container_ip, str)
                 endpoint = (container_ip, container_port)
                 log.info('Launched sibling container %s from image %s, listening on %s:%i',
                          container.name, image, container_ip, container_port)
             else:
-                ports = network_settings['Ports']
-                port = one(ports[f'{container_port}/tcp'])
+                while True:
+                    seconds = (datetime.now() - start).total_seconds()
+                    ports = container_info['NetworkSettings']['Ports'][f'{container_port}/tcp']
+                    if len(ports) > 0:
+                        break
+                    elif seconds > 3:
+                        raise RuntimeError('Unreachable TCP port', container_port, container.name)
+                    else:
+                        time.sleep(.33)
+                        container_info = cls._docker.api.inspect_container(container.id)
+                port = one(ports)
                 host_ip = port['HostIp']
                 host_port = int(port['HostPort'])
-                log.info('Launched container %s from image %s, '
+                log.info('Launched container %s from image %s after %.3fs, '
                          'with container port %s mapped to %s:%i on the host',
-                         container.name, image, container_port, host_ip, host_port)
+                         container.name, image, seconds, container_port, host_ip, host_port)
                 endpoint = (host_ip, host_port)
         except BaseException:  # no coverage
             container.kill()

